### ISSUE TITLE
Implement metadata schema

### ISSUE THREAD

**Issue by olafveerman** [energy-data/energydata.info#71]
Harmonize energydata.info metadata schema with development data hub.

Next step:

- get metadata schema @yanntanvez @Naigege

**Comment by olafveerman** [comment]
@carderne to provide list of fields, indicating:

- if they are mandatory
- type of field (pre-defined taxonomy)

**Comment by olafveerman** [comment]
From #105 : 

> See the updated metedata scheme:
[energydata Metadata_20170215.xlsx](https://github.com/energy-data/energydata.info/files/777879/energydata.Metadata_20170215.xlsx)
>
> A few notes:
> 1. All of the mandatory fields must be completed to continue (and add a red asterisk or something to indicate)
> 2. Data Type can default to Geospatial
> 3. Topics should be a lookup field allowing at least 1, at most 3 topics
> 4. At least one of Country _or_ Region should be entered, but both is allowed
> 5. Published Year: validate so just a year is entered (i.e. restrict to four digits between 1xxx - 2xxx)
>
> So from the current dataset creation page, the following can be removed: Version, Maintainer, Maintainer Email, Custom Fields

**Comment by carderne** [comment]
Sorry, I missed this topic. Please let me know if you want to discuss the metedata scheme, I'm not sure how self explanatory it is.

**Comment by olafveerman** [comment]
## New fields

| label | id | description | required | type |
| --- | --- | --- | --- | --- |
| Type | dataType | Identify type of dataset. | true | single select |
| Topics | topic | What does the dataset cover. | true | multi select |
| Region | region | The Bank Regions this dataset applies to. | true | multi select |
| Author | publisher | Author or TTL for datasets | false | text |
| Author email | contactEmail | Contact email for dataset questions | false | text |
| Published year | releaseDate | Publish date of source data | false | Date |
| Source  | source | Source from which dataset was obtained | false | text |
| Start date | startDate | Earliest data point in the dataset.| false | date |
| End date | endDate | Latest or most recent data point in the dataset | false | date |
| Map projection | mapProject | Name of the map projection | false | text |
| Status | status | The status of the data set | false | single select |

### Options

Type:
```
Geospatial
Microdata
Time series
Other
```

Topics:
```
Energy demand
Energy efficiency
Renewable energy
Energy access
Resource assessments
Measurement stations
Power system and utilities
Transmission and distribution
Pipeline networks
Physical features
Boundaries
Extractive industries
Project sites
Climate
Demographics
Surveys
Economics and prices
Policies and plans
Information technology
Environment
Transport
Water
Agriculture
Bioenergy
Geothermal
Hydropower
Solar
Wind
Thermal power
```

Region:
```
AFR Africa
EAP East Asia and Pacific
ECA Europe and Central Asia
LCR Latin America & Caribbean
SAR South Asia
MNA Middle East and North Africa
```

Status:
```
Complete
Obsolete
Ongoing 
Planned
Required
Under Development
```

## Changes to existing fields
The following fields should be made mandatory:

- title
- description
- country

**Comment by olafveerman** [comment]
Is this a fair summary @Naigege @carderne ?

cc @davidread

**Comment by Naigege** [comment]
@olafveerman Thanks! Below are the feedbacks from our side:

1. the existing "Tag" field will be removed and replaced by new "Topics" field. 
2. for the new "Topic" filed: we ask user to put at least 1 entry, and at most 3 entries from the list. 
3. for new "Region" filed and existing "Country" filed: user has to enter at least one. If "Country" is entered, "Region" will be not mandatory. Vice Versa. 
4. the existing "Visibility" filed can be removed, b/c we do not have other type. 
5. the new "type" field with options - geospatial/microdata/time series/others: please remove this, not required anymore. 
6. the existing fields - "Version", "Maintainer" and "Maintainer email": please remove. 
7. add a new filed - "Identifying Tag": a unique tag that can be entered to allow automatic scraping of a desired group of data-sets. 

In summary, the proposed fields are:
1. Title
2. Description
3. Topics
4. Country
5. Region
6. Organization
7. License
8. Source
9. Author
10. Author email
11. Published year
12. Start date
13. End date
14. Map projection
15. Status
16. Identifying tag 
17. User Custom Field/CKAN metadata integration plugin (we discussed this plugin yesterday)

Please let me know if you have any questions. 
Thanks!

**Comment by davidread** [comment]
> User Custom Field/CKAN metadata integration plugin (we discussed this plugin yesterday)

@olafveerman (or @Naigege) can you explain this field to me?

Do you mean doing something different with the existing custom fields?
<img width="522" alt="screen shot 2017-03-14 at 12 05 36" src="https://cloud.githubusercontent.com/assets/307612/23899826/9a14b1d2-08ae-11e7-81c5-5a5525251cff.png">

**Comment by davidread** [comment]
@olafveerman @Naigege 

You've asked for the 'Visibility' field to be removed. I'll simply hide it on the form, since it is needed for ckan's internal operation. But what do you want doing with these four datasets that are marked as private?
```
ckan=> select name from package where private=true and state='active';
        name
---------------------
 bla
 ferries-test-2
 zambia-power-plants
 distance-to-grid-1k
```

29 datasets have values for maintainer fields, which you've asked to be removed. Do these values need removing or kept?
```
select name, maintainer, maintainer_email from package where (maintainer <> '' or maintainer_email <> '') and state='active';
                                name                                |             maintainer              |        maintainer_email
--------------------------------------------------------------------+-------------------------------------+--------------------------------
 transmission-network-nigeria                                       | Dimitris Mentis                     | mentis@kth.se
 transmission-network-tanzania                                      | Dimitris Mentis                     | mentis@kth.se
 myanmar-existing-grid-medium-voltage-line-data                     | Shaky Sherpa                        |
 hydro-powerplants-in-rdc                                           | Cyril PERRET                        | c.perret@ied-sa.fr
 hydro-powerplants-in-central-africa                                | Cyril PERRET                        | c.perret@ied-sa.fr
 zambia-solar-mutanda-measurementdata-esmap                         | Oliver Knight                       | oknight@worldbank.org
 zambia-solar-lusaka-measurementdata-esmap                          | Oliver Knight                       | oknight@worldbank.org
 transmission-network-zambia                                        | Dimitris Mentis                     | mentis@kth.se
 myanmar-geospatial-based-least-cost-modeling-output                | Shaky Sherpa                        |
 kano-electricity-distribution-plc-kedco-mv-lines-2016              | Yann Tanvez                         | ytanvez@worldbank.org
 zambia-wind-measurement-data                                       | Oliver Knight                       | oknight@worldbank.org
 small-and-mini-hydropower-potential-in-sub-saharan-africa          | Alexandros Korkovelos               | alekor@kth.se
 zambia-solar-longe-measurementdata-esmap                           | Oliver Knight                       | oknight@worldbank.org
 zambia-solar-misamfu-measurementdata-esmap                         | Oliver Knight                       | oknight@worldbank.org
 zambia-solar-mochipapa-measurementdata-esmap                       | Oliver Knight                       | oknight@worldbank.org
 malawi-solar-kasungu-measurementdata-esmap                         | Oliver Knight                       | oknight@worldbank.org
 sub-saharan-africa-power-plant-2007                                |  Cecilia Brice<C3><B1>o-Garmendia         | cbricenogarmendi@worldbank.org
 malawi-solar-chileka-measurementdata-esmap                         | Oliver Knight                       | oknight@worldbank.org
 malawi-solar-mzuzu-measurementdata-esmap                           | Oliver Knight                       | oknight@worldbank.org
 zambia-solar-mountmakuluchilanga-measurementdata-esmap             | Oliver Knight                       | oknight@worldbank.org
 guinea-electricity-transmission-network-2014                       | Christopher Arderne                 | carderne@worldbank.org
 ssa-aicd-electricity-transmission-network                          | Cecilia Brice<C3><B1>o-Garmendia          | cbricenogarmendi@worldbank.org
 pakistan-transmission-network                                      | Clara Ivanescu                      | civanescu@worldbank.org
 africa-the-electicity-supply-system                                | Division of Energy Systems Analysis | admin@desa.kth.se
 energydata-info-user-manual                                        | Christopher Arderne                 | carderne@worldbank.org
 nigeria-kaduna-electricity-distribution-company-population-dataset | Yann Tanvez                         | ytanvez@worldbank.org
 rise-regulatory-indicators-for-sustainable-energy-2016             | Tanya Primiani                      | tprimiani@ifc.org
 mingov-zambia                                                      | Helga Treichel                      | htreichel@worldbank.org
 mingov-peru                                                        | Helga Treichel                      | htreichel@worldbank.org
(29 rows)
```

The tag field is well used. Is it definitely ok to simply delete this info?

The version field has not been used, so that's ok.

**Comment by olafveerman** [comment]
@davidread I removed the private datasets, these are testing artifacts.

I'll let @carderne and @Naigege get back to you on the `maintainer` field.

**Comment by carderne** [comment]
@davidread 
`Visibility`
I went to those four datasets and they appear to all be [deleted](https://energydata.info/dataset/distance-to-grid-1k)? Is that CKAN's internal way of deleting datasets? Anyway agree with just hiding the field.

`Maintainer`
Possible to automatically copy these over to `Author` in cases where that is blank? Otherwise I suspect these are mostly just repeats of what is in `Author`. Either way Naichen and I can manually check these afterwards.

`Tags`
Suggest not to delete it, but simply hide it from the user. If you could provide a csv dump of the current tags that might also be useful, as Naichen and I will be manually 're-topicing' everything.

**Comment by carderne** [comment]
@Naigege @davidread 
> User Custom Field/CKAN metadata integration plugin (we discussed this plugin yesterday)

I don't know what this is about.

**Comment by davidread** [comment]
'Map Projection' field - What's an example value for this? "Mercator"?

I'm wondering whether this is really what you want, because in other metadata catalogues, we've collected the "coordinate reference system" instead, e.g. "WGS84", so that we know on what ellipsoid the lat/long coordinates in the data refer too. I just want to check that's not what you're after here, so do confirm.

**Comment by carderne** [comment]
@davidread 
Good point. I don't know if you've been part of that conversation, but we're trying to maintain some level of compatibility with another WB project, DDH. I somehow missed it before but the key they're using is refSystem, so yes, please drop the `Map Projection` and include a CRS field!

**Comment by davidread** [comment]
@carderne Thanks for coming back on these so quickly!

`Visibility` - yep that's fine now Olaf deleted them. (And yes, ckan makes them still visible to admins, so that they can be undeleted if necessary.)

`Tags` - great, I'll simply hide them then. The dump of tags for your to help setting topics is here: https://gist.github.com/davidread/f6885b167f6ed5e3206c319c1055c43e

`Map Projection` - perfect, I'll reword it as CRS then. 'Coordinate Reference System:' e.g. 'WGS84' stored as 'ref-system' (CKAN's convention is kebab-case, rather than camelCase).

`Maintainer` - you're right about lots of duplication with the author fields. Ignoring those, here are the remaining ones where you'll need to decide what to do about the maintainer value:
```
> select name, author, author_email, maintainer, maintainer_email from package where (maintainer <> '' or maintainer_email <> '') and author <> maintainer and state='active';
                           name                            |     author      |        author_email        |         maintainer          |        maintainer_email
-----------------------------------------------------------+-----------------+----------------------------+-----------------------------+--------------------------------
 zambia-solar-mutanda-measurementdata-esmap                | ESMAP           | esmap@worldbank.org        | Oliver Knight               | oknight@worldbank.org
 zambia-solar-lusaka-measurementdata-esmap                 | ESMAP           | esmap@worldbank.org        | Oliver Knight               | oknight@worldbank.org
 myanmar-geospatial-based-least-cost-modeling-output       | shaky Sherpa    | ss3491@columbia.edu        | Shaky Sherpa                |
 zambia-wind-measurement-data                              | DNV GL          |  shant.dokouzian@dnvgl.com | Oliver Knight               | oknight@worldbank.org
 small-and-mini-hydropower-potential-in-sub-saharan-africa | Dimitris Mentis | mentis@kth.se              | Alexandros Korkovelos       | alekor@kth.se
 zambia-solar-longe-measurementdata-esmap                  | ESMAP           | esmap@worldbank.org        | Oliver Knight               | oknight@worldbank.org
 zambia-solar-misamfu-measurementdata-esmap                | ESMAP           | esmap@worldbank.org        | Oliver Knight               | oknight@worldbank.org
 zambia-solar-mochipapa-measurementdata-esmap              | ESMAP           | esmap@worldbank.org        | Oliver Knight               | oknight@worldbank.org
 malawi-solar-kasungu-measurementdata-esmap                | ESMAP           | esmap@worldbank.org        | Oliver Knight               | oknight@worldbank.org
 sub-saharan-africa-power-plant-2007                       |                 |                            |  Cecilia Brice<C3><B1>o-Garmendia | cbricenogarmendi@worldbank.org
 malawi-solar-chileka-measurementdata-esmap                | ESMAP           | esmap@worldbank.org        | Oliver Knight               | oknight@worldbank.org
 malawi-solar-mzuzu-measurementdata-esmap                  | ESMAP           | esmap@worldbank.org        | Oliver Knight               | oknight@worldbank.org
 zambia-solar-mountmakuluchilanga-measurementdata-esmap    | ESMAP           | esmap@worldbank.org        | Oliver Knight               | oknight@worldbank.org
 guinea-electricity-transmission-network-2014              |                 |                            | Christopher Arderne         | carderne@worldbank.org
 mingov-zambia                                             |                 |                            | Helga Treichel              | htreichel@worldbank.org
 mingov-peru                                               |                 |                            | Helga Treichel              | htreichel@worldbank.org
(16 rows)
```

**Comment by Naigege** [comment]
@davidread @carderne Thanks for these clarification! 

Maintainer - Let us use Maintainer and Maintainer's email for Author and Author email in our new metadata, for these datasets you list. Only for the dataset below, let us keep author and author email and abandon maintainer's  
small-and-mini-hydropower-potential-in-sub-saharan-africa | Dimitris Mentis | mentis@kth.se              | Alexandros Korkovelos       | alekor@kth.se

@olafveerman Could you help me explain User Custom Field/CKAN metadata integration plugin (we discussed this plugin yesterday) to @davidread ? Thanks!

**Comment by davidread** [comment]
@Naigege Ok, understood - thanks.

**Comment by olafveerman** [comment]
@Naigege 
We spoke about the 'CKAN metadata integration plugin' referring to the full set of changes that @davidread is currently making. I don't think that requires an extra field.

**Comment by davidread** [comment]
`Identifying tag` - Is it necessary to have this unique identifier field when we already have the unique name, as used in the URL?
<img width="426" alt="screen shot 2017-03-15 at 12 01 22" src="https://cloud.githubusercontent.com/assets/307612/23947648/8c420b44-0977-11e7-9bf5-8e8966744f52.png">
They are generated from the title automatically and ensured to be unique. However that would be no use to you if you wanted to go beyond the allowed characters: [a-z0-9-] or if these identifiers already existed so you just want to type them in. So it depends on your needs.

Also the word 'tag' could be confusing here, as that usually refers to keywords and related to taxonomies. But I understand naming it that if that keeps it in line with this other DDH project you mentioned.

**Comment by carderne** [comment]
@davidread 
We had a use case from some people we spoke to, that wanted to keep track of a group of datasets in an easy, machine-friendly way. So all datasets related to a specific project (say) could be given the same `Identifying tag` and a computer could easily look that up and harvest all the datasets that match.

If this can be achieved another way, and if you have suggestions for a better/more consistent name, I'm all ears.

**Comment by davidread** [comment]
@carderne Thanks, that really helps. How about calling the field "Group"? It could just be a simple text field which I add to the search index so that someone could find all the datasets in a group by searching for `group:"Wind Measurement for Developing Wind Power Plan"` for example.

**Comment by carderne** [comment]
@davidread Sounds good!

**Comment by carderne** [comment]
@davidread @olafveerman 
I noticed the 'Group' field is pre-populated? (In the case of the dataset I'm adding now `be8e5286-de7f-4759-9766-8b045e698798`).

Is this necessary? I foresee long strings of text and numbers confusing people that want to use that field.

**Comment by carderne** [comment]
@davidread 

It doesn't seem to matter which organization I use.
See a print out below:
[error.pdf](https://github.com/energy-data/energydata.info/files/855577/error.pdf)

**Comment by davidread** [comment]
Ah, got you, it's due to the parameter e.g. https://energydata.info/dataset/new?group=ee7925b8-b034-4a49-96e2-63013c9fcafc

I'll work out a fix

**Comment by carderne** [comment]
Ok. It's not a big issue, as the user can simply erase it and enter their own Group value. I just think less technically able users will be scared off by a random string of characters in the form...

**Comment by davidread** [comment]
@carderne I've done a fix for this bug and I'll leave it to @olafveerman to merge and deploy 
https://github.com/energy-data/energydata.info/pull/136

**Comment by olafveerman** [comment]
@carderne T.his was deployed today. Could you test again?

**Comment by carderne** [comment]
Looks good!

**Comment by carderne** [comment]
@davidread 
(**Wait until after the launch!**)
Add `Global` as an option for `Region`.

**Comment by davidread** [comment]
@carderne I'm not contracted any further on this - please ask @olafveerman in the first instance

**Comment by Naigege** [comment]
@olafveerman  Change published year as mandatory field. Then we are good.

**Comment by Naigege** [comment]
link to #185
